File: src/platform.rs
//! Platform detection utilities

use std::error::Error;

/// Detect the current platform and return the artifact name
///
/// # Errors
///
/// Returns an error if the platform is not supported
pub fn get_platform_artifact_name() -> Result<String, Box<dyn Error>> {
    let os = std::env::consts::OS;
    let arch = std::env::consts::ARCH;

    let platform_str = match (os, arch) {
        ("linux", "x86_64") => "linux-x64",
        ("linux", "aarch64") => "linux-arm64",
        ("macos", "x86_64") => "macos-x64",
        ("macos", "aarch64") => "macos-arm64",
        _ => {
            return Err(format!(
                "Unsupported platform: {} {}. Pre-built binaries are only available for:\n\
                 - Linux x64 (linux-x64)\n\
                 - Linux ARM64 (linux-arm64)\n\
                 - macOS x64 (macos-x64)\n\
                 - macOS ARM64 (macos-arm64)\n\
                 \n\
                 Consider using 'zircon build' to build from source instead.",
                os,
                arch
            )
            .into());
        }
    };

    Ok(format!("zrc-{}.tar.gz", platform_str))
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_get_platform_artifact_name() {
        // Just test that it doesn't panic
        let _ = super::get_platform_artifact_name();
    }
}

File: src/platform.c
#include "platform.h"
#include <stdio.h>

int detect_platform_and_arch(char *platform, size_t platform_size, char *arch, size_t arch_size)
{
#if defined(__APPLE__) && defined(_M_X64)
    fprintf(stderr, "Zirco no longer builds binaries for x64 macOS, please build from source.\n");
    return -1;
#endif

#ifdef __linux__
    snprintf(platform, platform_size, "linux");
#elif defined(__APPLE__)
    snprintf(platform, platform_size, "macos");
#else
    fprintf(stderr, "Unsupported platform, only Linux and macOS are supported.\n");
    return -1;
#endif

#if defined(__x86_64__) || defined(_M_X64)
    snprintf(arch, arch_size, "x64");
#elif defined(__aarch64__) || defined(_M_ARM64)
    snprintf(arch, arch_size, "arm64");
#else
    fprintf(stderr, "Unsupported architecture, only x64 and arm64 are supported.\n");
    return -1;
#endif

    return 0;
}
